File: mobile/lib/features/cv/domain/entities/cv.dart
import '../value_objects/cv_id.dart';
import '../value_objects/cv_style_ref.dart';
import 'certification.dart';
import 'copy_with_sentinel.dart';
import 'education.dart';
import 'experience.dart';
import 'language.dart';
import 'personal_info.dart';
import 'project.dart';
import 'skill.dart';

/// Entite racine du domaine CV : immuable, pure (aucune dependance Flutter,
/// HTTP ou JSON), avec collections copiees defensivement et non modifiables.
///
/// La (de)serialisation et le calcul de completion vivent hors de l'entite
/// (respectivement dans `data/` et `domain/policies/`).
final class CvEntity {
  final CvId? id;
  final String titre;
  final PersonalInfo? personalInfo;
  final List<Education> educations;
  final List<Experience> experiences;
  final List<Skill> skills;
  final List<Language> languages;
  final List<Certification> certifications;
  final List<Project> projects;
  final int viewCount;
  final int? parentCvId;
  final String? varianteLabel;
  final int? variantCount;
  final String? shareToken;
  final bool publicEnabled;
  final bool publicDownloadsEnabled;
  final bool publicContactEnabled;
  final int downloadCount;
  final int shareCount;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final CvStyleRef style;

  CvEntity({
    this.id,
    required this.titre,
    this.personalInfo,
    List<Education> educations = const [],
    List<Experience> experiences = const [],
    List<Skill> skills = const [],
    List<Language> languages = const [],
    List<Certification> certifications = const [],
    List<Project> projects = const [],
    this.viewCount = 0,
    this.parentCvId,
    this.varianteLabel,
    this.variantCount,
    this.shareToken,
    this.publicEnabled = false,
    this.publicDownloadsEnabled = false,
    this.publicContactEnabled = false,
    this.downloadCount = 0,
    this.shareCount = 0,
    this.createdAt,
    this.updatedAt,
    this.style = CvStyleRef.fallback,
  })  : educations = List.unmodifiable(educations),
        experiences = List.unmodifiable(experiences),
        skills = List.unmodifiable(skills),
        languages = List.unmodifiable(languages),
        certifications = List.unmodifiable(certifications),
        projects = List.unmodifiable(projects);

  /// True si ce CV est une variante adaptee a une offre.
  bool get isVariante => varianteLabel != null;

  CvEntity copyWith({
    Object? id = unsetSentinel,
    String? titre,
    Object? personalInfo = unsetSentinel,
    List<Education>? educations,
    List<Experience>? experiences,
    List<Skill>? skills,
    List<Language>? languages,
    List<Certification>? certifications,
    List<Project>? projects,
    int? viewCount,
    Object? parentCvId = unsetSentinel,
    Object? varianteLabel = unsetSentinel,
    Object? variantCount = unsetSentinel,
    Object? shareToken = unsetSentinel,
    bool? publicEnabled,
    bool? publicDownloadsEnabled,
    bool? publicContactEnabled,
    int? downloadCount,
    int? shareCount,
    Object? createdAt = unsetSentinel,
    Object? updatedAt = unsetSentinel,
    CvStyleRef? style,
  }) {
    return CvEntity(
      id: identical(id, unsetSentinel) ? this.id : id as CvId?,
      titre: titre ?? this.titre,
      personalInfo: identical(personalInfo, unsetSentinel)
          ? this.personalInfo
          : personalInfo as PersonalInfo?,
      educations: educations ?? this.educations,
      experiences: experiences ?? this.experiences,
      skills: skills ?? this.skills,
      languages: languages ?? this.languages,
      certifications: certifications ?? this.certifications,
      projects: projects ?? this.projects,
      viewCount: viewCount ?? this.viewCount,
      parentCvId: identical(parentCvId, unsetSentinel)
          ? this.parentCvId
          : parentCvId as int?,
      varianteLabel: identical(varianteLabel, unsetSentinel)
          ? this.varianteLabel
          : varianteLabel as String?,
      variantCount: identical(variantCount, unsetSentinel)
          ? this.variantCount
          : variantCount as int?,
      shareToken: identical(shareToken, unsetSentinel)
          ? this.shareToken
          : shareToken as String?,
      publicEnabled: publicEnabled ?? this.publicEnabled,
      publicDownloadsEnabled:
          publicDownloadsEnabled ?? this.publicDownloadsEnabled,
      publicContactEnabled: publicContactEnabled ?? this.publicContactEnabled,
      downloadCount: downloadCount ?? this.downloadCount,
      shareCount: shareCount ?? this.shareCount,
      createdAt: identical(createdAt, unsetSentinel)
          ? this.createdAt
          : createdAt as DateTime?,
      updatedAt: identical(updatedAt, unsetSentinel)
          ? this.updatedAt
          : updatedAt as DateTime?,
      style: style ?? this.style,
    );
  }

  @override
  bool operator ==(Object other) =>
      other is CvEntity &&
      other.id == id &&
      other.titre == titre &&
      other.personalInfo == personalInfo &&
      listEquals(other.educations, educations) &&
      listEquals(other.experiences, experiences) &&
      listEquals(other.skills, skills) &&
      listEquals(other.languages, languages) &&
      listEquals(other.certifications, certifications) &&
      listEquals(other.projects, projects) &&
      other.viewCount == viewCount &&
      other.parentCvId == parentCvId &&
      other.varianteLabel == varianteLabel &&
      other.variantCount == variantCount &&
      other.shareToken == shareToken &&
      other.publicEnabled == publicEnabled &&
      other.publicDownloadsEnabled == publicDownloadsEnabled &&
      other.publicContactEnabled == publicContactEnabled &&
      other.downloadCount == downloadCount &&
      other.shareCount == shareCount &&
      other.createdAt == createdAt &&
      other.updatedAt == updatedAt &&
      other.style == style;

  @override
  int get hashCode => Object.hashAll([
        id,
        titre,
        personalInfo,
        Object.hashAll(educations),
        Object.hashAll(experiences),
        Object.hashAll(skills),
        Object.hashAll(languages),
        Object.hashAll(certifications),
        Object.hashAll(projects),
        viewCount,
        parentCvId,
        varianteLabel,
        variantCount,
        shareToken,
        publicEnabled,
        publicDownloadsEnabled,
        publicContactEnabled,
        downloadCount,
        shareCount,
        createdAt,
        updatedAt,
        style,
      ]);
}

File: backend/src/main/java/com/cvmobile/cv/domain/model/Cv.java
package com.cvmobile.cv.domain.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Agregat de domaine representant un CV et ses sections.
 *
 * <p><em>Modele de domaine pur</em> : aucune dependance a Spring, JPA, Lombok
 * ou a un type web. Il protege ses invariants (titre et proprietaire
 * obligatoires, variante liee a un parent), n'expose ses collections que par
 * des vues non modifiables, et reference son proprietaire par le seul
 * {@code ownerId} — frontiere nette avec le module compte non encore migre.
 */
public final class Cv {

    private Long id;
    private String titre;
    private final long ownerId;
    private PersonalInfo personalInfo;
    private CvStyle style;

    private final List<Experience> experiences = new ArrayList<>();
    private final List<Education> educations = new ArrayList<>();
    private final List<Skill> skills = new ArrayList<>();
    private final List<Language> languages = new ArrayList<>();
    private final List<Certification> certifications = new ArrayList<>();
    private final List<Project> projects = new ArrayList<>();

    private final Long parentId;
    private final String varianteLabel;

    private int viewCount;
    private int downloadCount;
    private int shareCount;

    private Cv(Long id, String titre, long ownerId, CvStyle style,
               Long parentId, String varianteLabel) {
        this.id = id;
        this.titre = requireTitle(titre);
        this.ownerId = requirePositiveOwner(ownerId);
        this.style = requireStyle(style);
        this.parentId = parentId;
        this.varianteLabel = varianteLabel;
    }

    /**
     * Cree un CV original vierge avec le style par defaut.
     *
     * @throws IllegalArgumentException si le titre est vide ou le proprietaire
     *                                  invalide
     */
    public static Cv create(String titre, long ownerId) {
        return new Cv(null, titre, ownerId, CvStyle.defaults(), null, null);
    }

    /**
     * Cree une variante liee a un CV parent et etiquetee par une offre.
     *
     * @throws IllegalArgumentException si le parent est absent ou le label vide
     */
    public static Cv createVariant(String titre, long ownerId,
                                   Long parentId, String label) {
        if (parentId == null) {
            throw new IllegalArgumentException(
                    "Une variante doit referencer un CV parent (parentId).");
        }
        String safeLabel = DomainText.requireText(label, "label");
        return new Cv(null, titre, ownerId, CvStyle.defaults(), parentId, safeLabel);
    }

    /**
     * Reconstitue un agregat existant (utilise par la couche de persistance).
     * Reserve aux adaptateurs : ne pas appeler depuis la logique metier, qui
     * doit passer par {@link #create} ou {@link #createVariant}.
     */
    public static Cv rehydrate(Long id, String titre, long ownerId, CvStyle style,
                               Long parentId, String varianteLabel,
                               int viewCount, int downloadCount, int shareCount) {
        Cv cv = new Cv(id, titre, ownerId, style, parentId, varianteLabel);
        cv.viewCount = requireNonNegative(viewCount, "viewCount");
        cv.downloadCount = requireNonNegative(downloadCount, "downloadCount");
        cv.shareCount = requireNonNegative(shareCount, "shareCount");
        return cv;
    }

    // ── Identite et proprietes scalaires ─────────────────────────────

    public Long getId() { return id; }

    public String getTitre() { return titre; }

    public long getOwnerId() { return ownerId; }

    public PersonalInfo getPersonalInfo() { return personalInfo; }

    public CvStyle getStyle() { return style; }

    public Long getParentId() { return parentId; }

    public String getVarianteLabel() { return varianteLabel; }

    public int getViewCount() { return viewCount; }

    public int getDownloadCount() { return downloadCount; }

    public int getShareCount() { return shareCount; }

    /**
     * Affecte l'identifiant technique une fois le CV persiste. Reserve a la
     * couche de persistance; refuse toute reaffectation.
     */
    public void assignId(Long persistedId) {
        if (this.id != null) {
            throw new IllegalStateException(
                    "L'identifiant du CV est deja affecte et immuable.");
        }
        if (persistedId == null) {
            throw new IllegalArgumentException("persistedId ne peut etre null.");
        }
        this.id = persistedId;
    }

    /** Un CV est une variante des lors qu'il porte un label de variante. */
    public boolean isVariante() {
        return varianteLabel != null;
    }

    // ── Comportements metier ─────────────────────────────────────────

    /** Renomme le CV en normalisant et validant le titre. */
    public void rename(String newTitre) {
        this.titre = requireTitle(newTitre);
    }

    /** Remplace le style; refuse un style absent. */
    public void changeStyle(CvStyle newStyle) {
        this.style = requireStyle(newStyle);
    }

    /** Met a jour les informations personnelles (peut etre nul si effacees). */
    public void changePersonalInfo(PersonalInfo newPersonalInfo) {
        this.personalInfo = newPersonalInfo;
    }

    /** Enregistre une consultation (compteur calcule cote serveur). */
    public void registerView() {
        this.viewCount++;
    }

    /** Enregistre un telechargement (compteur calcule cote serveur). */
    public void registerDownload() {
        this.downloadCount++;
    }

    /** Enregistre un partage (compteur calcule cote serveur). */
    public void registerShare() {
        this.shareCount++;
    }

    // ── Collections : vues non modifiables + mutations controlees ─────

    public List<Experience> getExperiences() {
        return Collections.unmodifiableList(experiences);
    }

    public List<Education> getEducations() {
        return Collections.unmodifiableList(educations);
    }

    public List<Skill> getSkills() {
        return Collections.unmodifiableList(skills);
    }

    public List<Language> getLanguages() {
        return Collections.unmodifiableList(languages);
    }

    public List<Certification> getCertifications() {
        return Collections.unmodifiableList(certifications);
    }

    public List<Project> getProjects() {
        return Collections.unmodifiableList(projects);
    }

    public void addExperience(Experience e) {
        experiences.add(require(e, "experience"));
    }

    public void addEducation(Education e) {
        educations.add(require(e, "education"));
    }

    public void addSkill(Skill s) {
        skills.add(require(s, "skill"));
    }

    public void addLanguage(Language l) {
        languages.add(require(l, "language"));
    }

    public void addCertification(Certification c) {
        certifications.add(require(c, "certification"));
    }

    public void addProject(Project p) {
        projects.add(require(p, "project"));
    }

    public void replaceExperiences(List<Experience> items) {
        replace(experiences, items, "experience");
    }

    public void replaceEducations(List<Education> items) {
        replace(educations, items, "education");
    }

    public void replaceSkills(List<Skill> items) {
        replace(skills, items, "skill");
    }

    public void replaceLanguages(List<Language> items) {
        replace(languages, items, "language");
    }

    public void replaceCertifications(List<Certification> items) {
        replace(certifications, items, "certification");
    }

    public void replaceProjects(List<Project> items) {
        replace(projects, items, "project");
    }

    // ── Helpers prives ───────────────────────────────────────────────

    private static <T> void replace(List<T> target, List<T> items, String label) {
        target.clear();
        if (items == null) {
            return;
        }
        for (T item : items) {
            target.add(require(item, label));
        }
    }

    private static <T> T require(T value, String label) {
        if (value == null) {
            throw new IllegalArgumentException(
                    "Un element '" + label + "' ne peut etre null.");
        }
        return value;
    }

    private static String requireTitle(String titre) {
        return DomainText.requireText(titre, "titre");
    }

    private static CvStyle requireStyle(CvStyle style) {
        if (style == null) {
            throw new IllegalArgumentException("Le champ 'style' est obligatoire.");
        }
        return style;
    }

    private static long requirePositiveOwner(long ownerId) {
        if (ownerId <= 0) {
            throw new IllegalArgumentException(
                    "Le champ 'ownerId' doit etre strictement positif (recu : "
                            + ownerId + ").");
        }
        return ownerId;
    }

    private static int requireNonNegative(int value, String field) {
        if (value < 0) {
            throw new IllegalArgumentException(
                    "Le champ '" + field + "' ne peut etre negatif (recu : "
                            + value + ").");
        }
        return value;
    }
}
